File: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/dominio/irpf/faixas/FaixaFiscal.java
/*
 * Decompiled with CFR 0.152.
 * 
 * Could not load the following classes:
 *  javax.persistence.Column
 *  javax.persistence.MappedSuperclass
 */
package br.jus.trt8.pjecalc.negocio.dominio.irpf.faixas;

import br.jus.trt8.pjecalc.base.comum.Utils;
import br.jus.trt8.pjecalc.negocio.comum.MensagemDeRecurso;
import br.jus.trt8.pjecalc.negocio.comum.exceptions.NegocioException;
import br.jus.trt8.pjecalc.negocio.constantes.Mensagens;
import java.io.Serializable;
import java.math.BigDecimal;
import javax.persistence.Column;
import javax.persistence.MappedSuperclass;

@MappedSuperclass
public abstract class FaixaFiscal
implements Serializable {
    private static final long serialVersionUID = 746811403239854157L;
    @Column
    private BigDecimal valorInicial;
    @Column
    private BigDecimal valorFinal;
    @Column
    private BigDecimal aliquota;
    @Column
    private BigDecimal deducao;

    public FaixaFiscal() {
    }

    public FaixaFiscal(BigDecimal valorInicial, BigDecimal valorFinal, BigDecimal aliquota, BigDecimal deducao) {
        this.valorInicial = valorInicial;
        this.valorFinal = valorFinal;
        this.aliquota = aliquota;
        this.deducao = deducao;
    }

    public FaixaFiscal validar(String nomeTabela, NegocioException excecao, boolean validarValorFinal) {
        if (Utils.nulo(this.getValorInicial())) {
            excecao.adicionarMensagemDeRecurso(new MensagemDeRecurso("valorInicialFaixa" + this.getDiscriminador() + nomeTabela, Mensagens.MSG0003, "Valor Inicial da Faixa " + this.getDiscriminador()));
        }
        if (validarValorFinal && Utils.nulo(this.getValorFinal())) {
            excecao.adicionarMensagemDeRecurso(new MensagemDeRecurso("valorFinalFaixa" + this.getDiscriminador() + nomeTabela, Mensagens.MSG0003, "Valor Final da Faixa " + this.getDiscriminador()));
        }
        if (Utils.nulo(this.getAliquota())) {
            excecao.adicionarMensagemDeRecurso(new MensagemDeRecurso("aliquotaFaixa" + this.getDiscriminador() + nomeTabela, Mensagens.MSG0003, "Al\u00edquota da Faixa " + this.getDiscriminador()));
        }
        if (Utils.nulo(this.getDeducao())) {
            excecao.adicionarMensagemDeRecurso(new MensagemDeRecurso("deducaoFaixa" + this.getDiscriminador() + nomeTabela, Mensagens.MSG0003, "Parcela a Deduzir " + this.getDiscriminador()));
        }
        if (Utils.naoNulos(this.getValorFinal(), this.getValorInicial()) && this.isValorFinalMenorOuIgualQueValorInicial()) {
            excecao.adicionarMensagemDeRecurso(new MensagemDeRecurso("valorFinalFaixa" + this.getDiscriminador() + nomeTabela, Mensagens.MSG0007, "Valor Final", "Valor Inicial"));
        }
        return this;
    }

    public abstract String getDiscriminador();

    public boolean isValorFinalMenorOuIgualQueValorInicial() {
        return this.valorFinal.compareTo(this.valorInicial) <= 0;
    }

    public void adicionarNovoValorInicial(FaixaFiscal faixa) {
        if (Utils.naoNulo(faixa)) {
            this.valorInicial = faixa.getValorFinal().add(new BigDecimal("0.01"), Utils.CONTEXTO_MATEMATICO);
        }
    }

    public void sugerirValorMaximoParaValorFinal() {
        this.valorFinal = null;
    }

    public BigDecimal getValorInicial() {
        return this.valorInicial;
    }

    public void setValorInicial(BigDecimal valorInicial) {
        this.valorInicial = valorInicial;
    }

    public BigDecimal getValorFinal() {
        return this.valorFinal;
    }

    public void setValorFinal(BigDecimal valorFinal) {
        this.valorFinal = valorFinal;
    }

    public BigDecimal getAliquota() {
        return this.aliquota;
    }

    public void setAliquota(BigDecimal aliquota) {
        this.aliquota = aliquota;
    }

    public BigDecimal getDeducao() {
        return this.deducao;
    }

    public void setDeducao(BigDecimal deducao) {
        this.deducao = deducao;
    }
}

File: src/lib/pjecalc/core/dominio/irpf/faixas/faixa-fiscal.ts
/**
 * PJe-Calc v2.15.1 — FaixaFiscal (abstract) + 5 subclasses concretas
 * Porte 1:1 de:
 *   br.jus.trt8.pjecalc.negocio.dominio.irpf.faixas.FaixaFiscal
 *   + PrimeiraFaixaFiscal / SegundaFaixaFiscal / TerceiraFaixaFiscal /
 *     QuartaFaixaFiscal / QuintaFaixaFiscal
 *
 * Ref Java: pjecalc-fonte/.../irpf/faixas/ (6 arquivos, ~250 LOC total)
 *
 * FaixaFiscal é a base de uma faixa de tabela progressiva de IR:
 *   [valorInicial, valorFinal] — faixa de base tributável
 *   aliquota     — percentual (ex: 7.5 para 7,5%)
 *   deducao      — parcela a deduzir
 */
import Decimal from 'decimal.js';

export abstract class FaixaFiscal {
  private valorInicial: Decimal | null = null;
  private valorFinal: Decimal | null = null;
  private aliquota: Decimal | null = null;
  private deducao: Decimal | null = null;

  constructor(
    valorInicial?: Decimal | null,
    valorFinal?: Decimal | null,
    aliquota?: Decimal | null,
    deducao?: Decimal | null,
  ) {
    if (valorInicial !== undefined) this.valorInicial = valorInicial ?? null;
    if (valorFinal !== undefined) this.valorFinal = valorFinal ?? null;
    if (aliquota !== undefined) this.aliquota = aliquota ?? null;
    if (deducao !== undefined) this.deducao = deducao ?? null;
  }

  abstract getDiscriminador(): string;

  getValorInicial(): Decimal | null { return this.valorInicial; }
  setValorInicial(v: Decimal | null): void { this.valorInicial = v; }

  getValorFinal(): Decimal | null { return this.valorFinal; }
  setValorFinal(v: Decimal | null): void { this.valorFinal = v; }

  getAliquota(): Decimal { return this.aliquota ?? new Decimal(0); }
  setAliquota(v: Decimal | null): void { this.aliquota = v; }

  getDeducao(): Decimal | null { return this.deducao; }
  setDeducao(v: Decimal | null): void { this.deducao = v; }

  isValorFinalMenorOuIgualQueValorInicial(): boolean {
    if (!this.valorFinal || !this.valorInicial) return false;
    return this.valorFinal.lessThanOrEqualTo(this.valorInicial);
  }

  adicionarNovoValorInicial(faixa: FaixaFiscal | null): void {
    if (faixa && faixa.getValorFinal()) {
      this.valorInicial = faixa.getValorFinal()!.plus('0.01');
    }
  }
}

export class PrimeiraFaixaFiscal extends FaixaFiscal {
  getDiscriminador(): string { return '1'; }
}

export class SegundaFaixaFiscal extends FaixaFiscal {
  getDiscriminador(): string { return '2'; }
}

export class TerceiraFaixaFiscal extends FaixaFiscal {
  getDiscriminador(): string { return '3'; }
}

export class QuartaFaixaFiscal extends FaixaFiscal {
  getDiscriminador(): string { return '4'; }
}

export class QuintaFaixaFiscal extends FaixaFiscal {
  getDiscriminador(): string { return '5'; }
}
